src/wrapper.jl

```diff
@@ -0,0 +1,34 @@
+"""
+A SemanticWrapper determines the semantics of data that it wraps.
+Any subtype needs to implement function `unwrap(wrapper)` that returns the
+wrapped data.
+"""
+abstract type SemanticWrapper end
+
+"""
+Mask wraps a segmentation mask.
+"""
+struct Mask{T<:AbstractArray} <: SemanticWrapper
+    img::T
+
+    function Mask(img::T) where {T<:AbstractArray}
+        new{T}(img)
+    end
+end
```

Comment:
> (In the current code, even T(img) fails because it lacks Mask{T}(img:T) where {T<:AbstractArray} constructor)

This explicit inner constructor would tell Julia to not generate the default one, which is

```julia
function Mask{T}(img::T) where {T<:AbstractArray}
    new{T}(img)
end
```

```suggestion
struct Mask{AT<:AbstractArray} <: SemanticWrapper
    img::AT
end
Mask(img::AT) where AT<:AbstractArray = Mask{AT}(img)
```

Reply:
Maybe we don't even need the outer constructor at all?

Reply:
We do need it to construct the Mast, e.g., in `augment((img, Mask(mask)), pl)`

Reply:
But this also works:
```julia
julia> struct X{T}
           x::T
       end

julia> X(rand(3,3))
X{Matrix{Float64}}([0.2612806068891389 0.059315707573238896 0.6150185638418035; 0.2465880508484979 0.9856722987483633 0.07305366657204693; 0.5904516249308993 0.9808303184122573 0.8042661645164286])

julia> X{Matrix{Float64}}(rand(3,3))
X{Matrix{Float64}}([0.8469810746751445 0.7867887651387087 0.23515539548505093; 0.5869099501487944 0.4487894589360999 0.5984530764510496; 0.3818262964812178 0.6849713644656961 0.013309626688948528])

julia> X(rand(3,3) .> 0.5)
X{BitMatrix}(Bool[1 1 1; 1 0 1; 0 1 1])

julia> X{Matrix{Bool}}(rand(3,3) .> 0.5)
X{Matrix{Bool}}(Bool[1 0 0; 1 0 1; 0 1 1])
```
What am I missing?

Reply:
Oops, yeah. You're correct.

Just remove the inner construct and use the default ones that Julia provides should be okay.

Reply:
Alright. So now there is the last issue:
```julia
julia> m = Augmentor.Mask(trues(3,3))
Augmentor.Mask{BitMatrix}(Bool[1 1 1; 1 1 1; 1 1 1])

julia> T = typeof(m)
Augmentor.Mask{BitMatrix}

julia> T(rand(3,3))
ERROR: InexactError: Bool(0.32253992909116924)
Stacktrace:
 [1] Bool
   @ ./float.jl:106 [inlined]
 [2] convert
   @ ./number.jl:7 [inlined]
 [3] BitMatrix(A::Matrix{Float64})
   @ Base ./bitarray.jl:521
 [4] convert
   @ ./bitarray.jl:577 [inlined]
 [5] Augmentor.Mask{BitMatrix}(img::Matrix{Float64})
   @ Augmentor ~/julia/Augmentor.jl/src/wrapper.jl:12
 [6] top-level scope
   @ REPL[4]:1
```
I.e., if an operation changes the type of the image, then this error could pop up. I am not sure if we need to worry in practice.

Reply:
A good question. I think we can do something like

```julia
basetype(T) = Base.typename(T).wrapper

# basetype(typeof(m)) gives Mask instead of Mask{T}
basetype(typeof(m))(rand(3, 3))
```

This is a missing API in julia and is recently discussed in https://github.com/JuliaLang/julia/issues/35543